<html>
<head>
<title>test__threading_vs_settrace.py</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<style type="text/css">
.s0 { color: #cc7832;}
.s1 { color: #a9b7c6;}
.s2 { color: #6a8759;}
.s3 { color: #808080;}
.s4 { color: #6897bb;}
</style>
</head>
<body bgcolor="#2b2b2b">
<table CELLSPACING=0 CELLPADDING=5 COLS=1 WIDTH="100%" BGCOLOR="#606060" >
<tr><td><center>
<font face="Arial, Helvetica" color="#000000">
test__threading_vs_settrace.py</font>
</center></td></tr></table>
<pre><span class="s0">from </span><span class="s1">__future__ </span><span class="s0">import </span><span class="s1">print_function</span>
<span class="s0">import </span><span class="s1">sys</span>
<span class="s0">import </span><span class="s1">subprocess</span>
<span class="s0">import </span><span class="s1">unittest</span>
<span class="s0">from </span><span class="s1">gevent.thread </span><span class="s0">import </span><span class="s1">allocate_lock</span>
<span class="s0">import </span><span class="s1">gevent.testing </span><span class="s0">as </span><span class="s1">greentest</span>

<span class="s1">script = </span><span class="s2">&quot;&quot;&quot; 
from gevent import monkey 
monkey.patch_all() # pragma: testrunner-no-monkey-combine 
import sys, os, threading, time 
 
 
# A deadlock-killer, to prevent the 
# testsuite to hang forever 
def killer(): 
    time.sleep(0.2) 
    sys.stdout.write('..program blocked; aborting!') 
    sys.stdout.flush() 
    os._exit(2) 
t = threading.Thread(target=killer) 
t.daemon = True 
t.start() 
 
 
def trace(frame, event, arg): 
    if threading is not None: 
        threading.current_thread() 
    return trace 
 
 
def doit(): 
    sys.stdout.write(&quot;..thread started..&quot;) 
 
 
def test1(): 
    t = threading.Thread(target=doit) 
    t.start() 
    t.join() 
    sys.settrace(None) 
 
sys.settrace(trace) 
if len(sys.argv) &gt; 1: 
    test1() 
 
sys.stdout.write(&quot;..finishing..&quot;) 
&quot;&quot;&quot;</span>


<span class="s0">class </span><span class="s1">TestTrace(unittest.TestCase):</span>
    <span class="s1">@greentest.skipOnPurePython(</span><span class="s2">&quot;Locks can be traced in Pure Python&quot;</span><span class="s1">)</span>
    <span class="s0">def </span><span class="s1">test_untraceable_lock(self):</span>
        <span class="s3"># Untraceable locks were part of the solution to https://bugs.python.org/issue1733757</span>
        <span class="s3"># which details a deadlock that could happen if a trace function invoked</span>
        <span class="s3"># threading.currentThread at shutdown time---the cleanup lock would be held</span>
        <span class="s3"># by the VM, and calling currentThread would try to acquire it again. The interpreter</span>
        <span class="s3"># changed in 2.6 to use the `with` statement (https://hg.python.org/cpython/rev/76f577a9ec03/),</span>
        <span class="s3"># which apparently doesn't trace in quite the same way.</span>
        <span class="s0">if </span><span class="s1">hasattr(sys</span><span class="s0">, </span><span class="s2">'gettrace'</span><span class="s1">):</span>
            <span class="s1">old = sys.gettrace()</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s1">old = </span><span class="s0">None</span>

        <span class="s1">lst = []</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">def </span><span class="s1">trace(frame</span><span class="s0">, </span><span class="s1">ev</span><span class="s0">, </span><span class="s1">_arg):</span>
                <span class="s1">lst.append((frame.f_code.co_filename</span><span class="s0">, </span><span class="s1">frame.f_lineno</span><span class="s0">, </span><span class="s1">ev))</span>
                <span class="s1">print(</span><span class="s2">&quot;TRACE: %s:%s %s&quot; </span><span class="s1">% lst[-</span><span class="s4">1</span><span class="s1">])</span>
                <span class="s0">return </span><span class="s1">trace</span>

            <span class="s0">with </span><span class="s1">allocate_lock():</span>
                <span class="s1">sys.settrace(trace)</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s1">sys.settrace(old)</span>

        <span class="s1">self.assertEqual(lst</span><span class="s0">, </span><span class="s1">[]</span><span class="s0">, </span><span class="s2">&quot;trace not empty&quot;</span><span class="s1">)</span>

    <span class="s1">@greentest.skipOnPurePython(</span><span class="s2">&quot;Locks can be traced in Pure Python&quot;</span><span class="s1">)</span>
    <span class="s0">def </span><span class="s1">test_untraceable_lock_uses_different_lock(self):</span>
        <span class="s0">if </span><span class="s1">hasattr(sys</span><span class="s0">, </span><span class="s2">'gettrace'</span><span class="s1">):</span>
            <span class="s1">old = sys.gettrace()</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s1">old = </span><span class="s0">None</span>

        <span class="s1">lst = []</span>
        <span class="s3"># we should be able to use unrelated locks from within the trace function</span>
        <span class="s1">l = allocate_lock()</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">def </span><span class="s1">trace(frame</span><span class="s0">, </span><span class="s1">ev</span><span class="s0">, </span><span class="s1">_arg):</span>
                <span class="s0">with </span><span class="s1">l:</span>
                    <span class="s1">lst.append((frame.f_code.co_filename</span><span class="s0">, </span><span class="s1">frame.f_lineno</span><span class="s0">, </span><span class="s1">ev))</span>
                <span class="s3"># print(&quot;TRACE: %s:%s %s&quot; % lst[-1])</span>
                <span class="s0">return </span><span class="s1">trace</span>

            <span class="s1">l2 = allocate_lock()</span>
            <span class="s1">sys.settrace(trace)</span>
            <span class="s3"># Separate functions, not the C-implemented `with` so the trace</span>
            <span class="s3"># function gets a crack at them</span>
            <span class="s1">l2.acquire()</span>
            <span class="s1">l2.release()</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s1">sys.settrace(old)</span>

        <span class="s3"># Have an assert so that we know if we miscompile</span>
        <span class="s1">self.assertTrue(lst</span><span class="s0">, </span><span class="s2">&quot;should not compile on pypy&quot;</span><span class="s1">)</span>

    <span class="s1">@greentest.skipOnPurePython(</span><span class="s2">&quot;Locks can be traced in Pure Python&quot;</span><span class="s1">)</span>
    <span class="s0">def </span><span class="s1">test_untraceable_lock_uses_same_lock(self):</span>
        <span class="s0">from </span><span class="s1">gevent.hub </span><span class="s0">import </span><span class="s1">LoopExit</span>
        <span class="s0">if </span><span class="s1">hasattr(sys</span><span class="s0">, </span><span class="s2">'gettrace'</span><span class="s1">):</span>
            <span class="s1">old = sys.gettrace()</span>
        <span class="s0">else</span><span class="s1">:</span>
            <span class="s1">old = </span><span class="s0">None</span>

        <span class="s1">lst = []</span>
        <span class="s1">e = </span><span class="s0">None</span>
        <span class="s3"># we should not be able to use the same lock from within the trace function</span>
        <span class="s3"># because it's over acquired but instead of deadlocking it raises an exception</span>
        <span class="s1">l = allocate_lock()</span>
        <span class="s0">try</span><span class="s1">:</span>
            <span class="s0">def </span><span class="s1">trace(frame</span><span class="s0">, </span><span class="s1">ev</span><span class="s0">, </span><span class="s1">_arg):</span>
                <span class="s0">with </span><span class="s1">l:</span>
                    <span class="s1">lst.append((frame.f_code.co_filename</span><span class="s0">, </span><span class="s1">frame.f_lineno</span><span class="s0">, </span><span class="s1">ev))</span>
                <span class="s0">return </span><span class="s1">trace</span>

            <span class="s1">sys.settrace(trace)</span>
            <span class="s3"># Separate functions, not the C-implemented `with` so the trace</span>
            <span class="s3"># function gets a crack at them</span>
            <span class="s1">l.acquire()</span>
        <span class="s0">except </span><span class="s1">LoopExit </span><span class="s0">as </span><span class="s1">ex:</span>
            <span class="s1">e = ex</span>
        <span class="s0">finally</span><span class="s1">:</span>
            <span class="s1">sys.settrace(old)</span>

        <span class="s3"># Have an assert so that we know if we miscompile</span>
        <span class="s1">self.assertTrue(lst</span><span class="s0">, </span><span class="s2">&quot;should not compile on pypy&quot;</span><span class="s1">)</span>
        <span class="s1">self.assertTrue(isinstance(e</span><span class="s0">, </span><span class="s1">LoopExit))</span>

    <span class="s0">def </span><span class="s1">run_script(self</span><span class="s0">, </span><span class="s1">more_args=()):</span>
        <span class="s0">if </span><span class="s1">(</span>
                <span class="s1">greentest.PYPY3</span>
                <span class="s0">and </span><span class="s1">greentest.RUNNING_ON_APPVEYOR</span>
                <span class="s0">and </span><span class="s1">sys.version_info[:</span><span class="s4">2</span><span class="s1">] == (</span><span class="s4">3</span><span class="s0">, </span><span class="s4">7</span><span class="s1">)</span>
        <span class="s1">):</span>
            <span class="s3"># Somehow launching the subprocess fails with exit code 1, and</span>
            <span class="s3"># produces no output. It's not clear why.</span>
            <span class="s1">self.skipTest(</span><span class="s2">&quot;Known to hang on AppVeyor&quot;</span><span class="s1">)</span>
        <span class="s1">args = [sys.executable</span><span class="s0">, </span><span class="s2">&quot;-u&quot;</span><span class="s0">, </span><span class="s2">&quot;-c&quot;</span><span class="s0">, </span><span class="s1">script]</span>
        <span class="s1">args.extend(more_args)</span>
        <span class="s1">rc = subprocess.call(args)</span>
        <span class="s1">self.assertNotEqual(rc</span><span class="s0">, </span><span class="s4">2</span><span class="s0">, </span><span class="s2">&quot;interpreter was blocked&quot;</span><span class="s1">)</span>
        <span class="s1">self.assertEqual(rc</span><span class="s0">, </span><span class="s4">0</span><span class="s0">, </span><span class="s2">&quot;Unexpected error&quot;</span><span class="s1">)</span>

    <span class="s0">def </span><span class="s1">test_finalize_with_trace(self):</span>
        <span class="s1">self.run_script()</span>

    <span class="s0">def </span><span class="s1">test_bootstrap_inner_with_trace(self):</span>
        <span class="s1">self.run_script([</span><span class="s2">&quot;1&quot;</span><span class="s1">])</span>


<span class="s0">if </span><span class="s1">__name__ == </span><span class="s2">&quot;__main__&quot;</span><span class="s1">:</span>
    <span class="s1">greentest.main()</span>
</pre>
</body>
</html>